### PR TITLE
Add user interaction tracking hooks and events

### PR DESCRIPTION
**Pull request by DaoDaoNoCode**
**Fixes:**
https://issues.redhat.com/browse/RHODS-659
https://issues.redhat.com/browse/RHODS-862
https://issues.redhat.com/browse/RHODS-864

**Analysis / Root cause:**
We need to add user interaction tracking hooks to use [segment.io](https://segment.com/) to track user interactions.

**Implementation:**
- [x] Create a function named `FireTrackingEvent`, it accepts 2 arguments `eventType` and `properties` (optional). We could use this function anywhere we want to fire events and send it to [segment.io](https://segment.com/)
- [x] Automatically fire a page event whenever the URL changes
- [x] For every enabled application, fire an event to indicate that the application is enabled (https://issues.redhat.com/browse/RHODS-862)
- [x] For every resource on the Resource page, fire an event when the user clicks on the resource card's link button (https://issues.redhat.com/browse/RHODS-864)
- [x] Only enable the user interaction tracking when there is a `segmentKey`, `username`, and `clusterID`
- [x] Fire an identify event when the user tracking function is enabled, the identifier is the `SHA-1` encoded `username`
- [x] Give the access of `ClusterVersion` resource to `odh-dashboard` `ClusterRole`, because we can only get the `clusterID` from it and send the `clusterID` along with every event

**Note:**
To opt-in the `segmentKey`, create a `Secret` resource named `rhods-segment-key`, set the key as `segmentKey` and the value as [your segment key]. After you create it and provide the correct segment key, odh-dashboard will automatically start user interaction tracking. To opt-out, just delete the `Secret` resource you created.

### CODE REVIEW

File: backend/src/routes/api/status/index.ts [hunk 1]
```diff
@@ -9,8 +9,15 @@ const status = async (
...
-  const { currentContext, namespace, currentUser } = fastify.kube;
-  const userName = request.headers['x-forwarded-user'] ?? currentUser.username;
+  const { currentContext, namespace, currentUser, customObjectsApi } = fastify.kube;
+  const userName = request.headers['x-forwarded-user'] ?? currentUser.name;
+  const clusterVersion = await customObjectsApi.getClusterCustomObject(
+    'config.openshift.io',
+    'v1',
+    'clusterversions',
+    'version',
+  );
```
Comment (jeff-phillips-18) [inline backend/src/routes/api/status/index.ts:19]: This should be done once at startup in `kube.fp` and added to the kube decorator.

File: frontend/src/app/App.tsx [hunk 2]
```diff
@@ -9,11 +9,19 @@ import Routes from './Routes';
 import NavSidebar from './NavSidebar';
 
 import './App.scss';
+import { useHistory } from 'react-router';
+import { useSegmentIOTracking } from '../utilities/useSegmentIOTracking';
+import { fireTrackingEvent, initSegment } from '../utilities/segmentIOUtils';
+import { RootState } from '../redux/types';
```
Comment (jeff-phillips-18) [inline frontend/src/app/App.tsx:15]: nit: as convention, we import outer modules first (react-router), then local modules, then CSS. Can't find a lint rule to support that though.

File: frontend/src/components/OdhDocCard.tsx [hunk 3]
```diff
@@ -15,13 +15,22 @@ import DocCardBadges from './DocCardBadges';
 import { makeCardVisible } from '../utilities/utils';
 
 import './OdhCard.scss';
+import { fireTrackingEvent } from '../utilities/segmentIOUtils';
```
Comment (jeff-phillips-18) [inline frontend/src/components/OdhDocCard.tsx:18]: same nit

File: frontend/src/components/OdhDocCard.tsx [hunk 4]
```diff
@@ -67,6 +76,7 @@ const OdhDocCard: React.FC<OdhDocCardProps> = ({ odhDoc, favorite, updateFavorit
     e.preventDefault();
     launchQuickStart(odhDoc.metadata.name, qsContext);
     makeCardVisible(odhDoc.metadata.name);
+    fireResourceAccessedEvent(odhDoc.metadata.name, odhDoc.metadata.type)();
```
Comment (jeff-phillips-18) [inline frontend/src/components/OdhDocCard.tsx:79]: Do we care if it is a `start` vs `continue` vs `restart`?

File: frontend/src/pages/enabledApplications/EnabledApplications.tsx [hunk 5]
```diff
@@ -8,6 +8,7 @@ import OdhAppCard from '../../components/OdhAppCard';
 import QuickStarts from '../../app/QuickStarts';
 
 import './EnabledApplications.scss';
+import { fireTrackingEvent } from '../../utilities/segmentIOUtils';
```
Comment (jeff-phillips-18) [inline frontend/src/pages/enabledApplications/EnabledApplications.tsx:11]: same nit

File: frontend/src/pages/enabledApplications/EnabledApplications.tsx [hunk 6]
```diff
@@ -20,6 +21,16 @@ type EnabledApplicationsInnerProps = {
...
+
+    // fire an individual segment.io tracking event for every enabled application
+    if (!isEmpty) {
+      components.forEach((c) => {
+        fireTrackingEvent('Application Enabled', {
+          name: c.metadata.name,
+        });
+      });
+    }
+
```
Comment (jeff-phillips-18) [inline frontend/src/pages/enabledApplications/EnabledApplications.tsx:33]: How often do we want this event? I would think only once so I would put this in a `React.useEffect` and only update  on `isEmpty` change.

File: frontend/src/pages/learningCenter/LearningCenterFilters.tsx [hunk 7]
```diff
@@ -34,6 +34,7 @@ import {
 } from './learningCenterUtils';
 
 import './LearningCenterFilter.scss';
+import { fireTrackingEvent } from '../../utilities/segmentIOUtils';
```
Comment (jeff-phillips-18) [inline frontend/src/pages/learningCenter/LearningCenterFilters.tsx:37]: same nit

File: frontend/src/utilities/segmentIOUtils.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+};
+
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export const initSegment = async (props) => {
+  const { segmentKey, username } = props;
+  const analytics = (window.analytics = window.analytics || []);
+  if (analytics.initialize) {
+    return;
+  }
+  if (analytics.invoked)
```
Comment (jeff-phillips-18) [inline frontend/src/utilities/segmentIOUtils.tsx:29]: brackets please

File: frontend/src/utilities/useSegmentIOTracking.tsx [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    return () => {
+      if (watchHandle) {
+        clearTimeout(watchHandle);
+      }
+    };
+    // Don't update when components are updated
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const retSegmentKey = useDeepCompareMemoize<string>(segmentKey);
```
Comment (jeff-phillips-18) [inline frontend/src/utilities/useSegmentIOTracking.tsx:41]: I don't think you need to memoize a string

File: package.json [hunk 10]
```diff
@@ -23,7 +23,7 @@
     "dev": "run-p -l dev:*",
     "dev:backend": "cd ./backend && npm run start:dev",
     "dev:frontend": "cd ./frontend && npm run start:dev",
-    "format": "prettier --write \"backend/**/*.js\" \"frontend/**/*.ts\" \"frontend/**/*.tsx\"",
+    "format": "prettier --write \"backend/**/*.ts\" \"frontend/**/*.ts\" \"frontend/**/*.tsx\"",
```
Comment (jeff-phillips-18) [inline package.json:26]: 👍

File: frontend/src/utilities/segmentIOUtils.tsx [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+      default:
+        window.analytics.track(eventType, { ...properties, clusterID });
+    }
+  }
+};
+
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export const initSegment = async (props) => {
+  const { segmentKey, username } = props;
+  const analytics = (window.analytics = window.analytics || []);
```
Comment (jeff-phillips-18) [inline frontend/src/utilities/segmentIOUtils.tsx:25]:
```suggestion
  const analytics = (window.analytics = window.analytics || {});
```

File: frontend/src/app/App.tsx [hunk 12]
```diff
@@ -27,12 +35,38 @@ const App: React.FC = () => {
     setIsNavOpen(!isNavOpen);
   };
 
+  // when there is a segment key and username and cluster id, initialize segment.io script
+  React.useEffect(() => {
+    if (segmentKey && loaded && !loadError && username && clusterID) {
+      window.clusterID = clusterID;
+      initSegment({ segmentKey, username });
```
Comment (jeff-phillips-18) [inline frontend/src/app/App.tsx:42]: Could we have `initSegment` return the `analytics` and store with redux rather than using `window`?

**Review by jeff-phillips-18 — CHANGES_REQUESTED** [review]
If we can use `redux` rather than `window` to store the analytics, that'd be better.

**Comment by jeff-phillips-18** [comment]
please squash the commits (small commits are good if the comment describes the change 😉  )

File: frontend/src/components/OdhDocCard.tsx (hunk 4 above)
Comment (DaoDaoNoCode) [reply frontend/src/components/OdhDocCard.tsx:79]: I don't see this in the JIRA issue. Hi, @anishasthana do we care about whether the user starts the quick start, or continues, or restarts? Do we also need to send that information?

File: frontend/src/pages/enabledApplications/EnabledApplications.tsx (hunk 6 above)
Comment (DaoDaoNoCode) [reply frontend/src/pages/enabledApplications/EnabledApplications.tsx:33]: @jeff-phillips-18 What if there is already an enabled application and the user enables another one? Should we use `components` as the dependency?

File: frontend/src/utilities/segmentIOUtils.tsx (hunk 11 above)
Comment (DaoDaoNoCode) [reply frontend/src/utilities/segmentIOUtils.tsx:25]: It has to be an array, but not an object. I don't know why but it seems this is how the segment.io script was designed.

File: frontend/src/app/App.tsx (hunk 12 above)
Comment (DaoDaoNoCode) [reply frontend/src/app/App.tsx:42]: @jeff-phillips-18 If we use redux, will it be difficult to be called? We cannot use redux hook functions inside the `fireTrackingEvent` function, so every time we call it we need to pass it through the arguments. Do you have any suggestions about how to avoid this situation if we want to use redux?

File: frontend/src/pages/enabledApplications/EnabledApplications.tsx (hunk 6 above)
Comment (jeff-phillips-18) [reply frontend/src/pages/enabledApplications/EnabledApplications.tsx:33]: Correct, but not on every render. We could do it instead in the resource watcher 🤷‍♂️

File: frontend/src/app/App.tsx (hunk 12 above)
Comment (jeff-phillips-18) [reply frontend/src/app/App.tsx:42]: I think it is ok to pass the analytics to the `fireTrackingEvent` function.

File: backend/src/types.ts [hunk 13]
```diff
@@ -30,6 +30,14 @@ export type K8sResourceCommon = {
...
 };
 
+// Minimal type for ClusterVersion
+export type ClusterVersionType = {
+  spec: {
+    clusterID: string;
+    channel?: string;
+  };
+} & K8sResourceCommon;
+
```
Comment (jeff-phillips-18) [inline backend/src/types.ts:40]: I don't think we need to add this. Lets just do it inline.

File: backend/src/plugins/kube.ts [hunk 14]
```diff
@@ -23,6 +24,22 @@ export default fp(async (fastify: FastifyInstance) => {
...
 
+  let clusterID;
+  try {
+    const clusterVersion = await customObjectsApi.getClusterCustomObject(
+      'config.openshift.io',
+      'v1',
+      'clusterversions',
+      'version',
+    );
+    clusterID = (clusterVersion.body as ClusterVersionType).spec.clusterID;
```
Comment (jeff-phillips-18) [inline backend/src/plugins/kube.ts:35]:
```suggestion
    clusterID = (clusterVersion.body as { spec: { clusterID: string; };}).spec.clusterID;
```

File: frontend/src/components/OdhDocCard.tsx [hunk 15]
```diff
@@ -22,6 +23,22 @@ type OdhDocCardProps = {
...
+  const quickStartLabel = getQuickStartLabel(name, qsContext);
+  fireTrackingEvent(
+    type === ODHDocType.QuickStart ? `Resource ${quickStartLabel}` : 'Resource Accessed',
+    {
+      name: name,
+      type: type,
+    },
+  );
+};
+
```
Comment (jeff-phillips-18) [inline frontend/src/components/OdhDocCard.tsx:41]: This should not be a global constant, define inside the component.

File: frontend/src/utilities/useWatchSegmentKey.tsx [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+      watchHandle = setTimeout(watchSegmentKey, POLL_INTERVAL);
+    };
+    watchSegmentKey();
+
+    return () => {
+      if (watchHandle) {
+        clearTimeout(watchHandle);
+      }
+    };
+    // Don't update when components are updated
```
Comment (jeff-phillips-18) [inline frontend/src/utilities/useWatchSegmentKey.tsx:36]: copy/paste error

**Review by jeff-phillips-18 — CHANGES_REQUESTED** [review]

File: frontend/src/utilities/useWatchDocs.tsx [hunk 17]
```diff
@@ -20,9 +20,9 @@ export const useWatchDocs = (
     const watchQuickStarts = () => {
       fetchDocs(docType)
         .then((updatedDocs: ODHDoc[]) => {
+          setDocs(updatedDocs);
           setLoaded(true);
           setLoadError(undefined);
-          setDocs(updatedDocs);
```
Comment (jeff-phillips-18) [inline frontend/src/utilities/useWatchDocs.tsx:25]: why this change?

File: frontend/src/pages/enabledApplications/EnabledApplications.tsx [hunk 18]
```diff
@@ -54,6 +60,24 @@ const EnabledApplications: React.FC = () => {
     );
   }, [components]);
 
+  React.useEffect(() => {
+    /*
+     * compare the current enabled applications and new fetched enabled applications
+     * fire an individual segment.io tracking event for every different enabled application
+     */
+    if (loaded) {
```
Comment (jeff-phillips-18) [inline frontend/src/pages/enabledApplications/EnabledApplications.tsx:68]:
```suggestion
    if (loaded && components.length) {
```

**Review by jeff-phillips-18 — APPROVED** [review]
